src/core/elf.d

```diff
@@ -0,0 +1,341 @@
+/**
+ * This code simplifies working with ELF binaries, e.g., iterating
+ * over loaded shared objects & their segments as well as reading
+ * headers and sections from ELF files.
+ *
+ * Reference: http://www.dwarfstd.org/
+ *
+ * Copyright: Copyright Digital Mars 2015 - 2015.
+ * License:   $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+ * Authors:   Yazan Dabain, Martin Kinkelin
+ * Source: $(DRUNTIMESRC core/elf.d)
+ */
+
+module core.elf;
+
+version (linux)             version = linux_or_bsd;
+else version (FreeBSD)      version = linux_or_bsd;
+else version (DragonFlyBSD) version = linux_or_bsd;
+
+version (linux_or_bsd):
+
+import core.sys.posix.fcntl;
+import core.sys.posix.unistd;
+
+version (linux)
+{
+    import core.sys.linux.link;
+    import core.sys.linux.elf;
+}
+else version (FreeBSD)
+{
+    import core.sys.freebsd.sys.link_elf;
+    import core.sys.freebsd.sys.elf;
+}
+else version (DragonFlyBSD)
+{
+    import core.sys.dragonflybsd.sys.link_elf;
+    import core.sys.dragonflybsd.sys.elf;
+}
+
+alias Elf_Phdr = ElfW!"Phdr";
+alias Elf_Ehdr = ElfW!"Ehdr";
+alias Elf_Shdr = ElfW!"Shdr";
+
+/****
+ * Enables iterating over the process' currently loaded shared objects.
+ */
+struct SharedObjects
+{
+@nogc nothrow:
+    alias Callback = int delegate(SharedObject);
+
+    static int opApply(scope Callback dg)
+    {
+        extern(C) int nativeCallback(dl_phdr_info* info, size_t, void* data)
+        {
+            auto dg = *cast(Callback*) data;
+            return dg(SharedObject(*info));
+        }
+
+        return dl_iterate_phdr(&nativeCallback, &dg);
+    }
+}
+
+struct SharedObject
+{
+@nogc nothrow:
+    /****
+     * Finds the shared object containing the specified address in one of its segments.
+     */
+    static bool findForAddress(in void* address, out SharedObject result)
+    {
+        version (linux)       enum IterateManually = true;
+        else version (NetBSD) enum IterateManually = true;
+        else                  enum IterateManually = false;
+
+        static if (IterateManually)
+        {
+            foreach (object; SharedObjects)
+            {
+                const(Elf_Phdr)* segment;
+                if (object.findSegmentForAddress(address, segment))
+                {
+                    result = object;
+                    return true;
+                }
+            }
+            return false;
+        }
+        else
+        {
+            return !!_rtld_addr_phdr(address, &result.info);
+        }
+    }
+
+    dl_phdr_info info;
```

Comment:
Do we want this to be directly exposed?

Reply:
I think so; `rt.sections_elf_backtrace` currently needs it for the TLS stuff, and I don't want to abstract it fully away, particularly since the struct is OS-dependent.